Objective-C/CBLDatabase.mm

```diff
@@ -722,45 +726,131 @@ - (nullable CBLScope*) scopeWithName: (NSString*)name error: (NSError**)error {
+                return nil;
+            }
+            
+            [collections addObject: c];
+        }
+        FLArray_Release(list);
+        
+        return [NSArray arrayWithArray: collections];
+    }
 }
 
 - (nullable CBLCollection*) createCollectionWithName: (NSString*)name
-                                      scope: (nullable NSString*)scope
-                                      error: (NSError**)error {
+                                               scope: (nullable NSString*)scopeName
+                                               error: (NSError**)error {
+    scopeName = scopeName ?: kCBLDefaultScopeName;
+    CBLStringBytes cName(name);
+    CBLStringBytes sName(scopeName);
+    C4CollectionSpec spec = { .name = cName, .scope = sName };
     
-    // TODO: add implementation
+    C4Collection* c4collection;
+    CBL_LOCK(self) {
+        if (![self mustBeOpen: error])
+            return nil;
+        
+        C4Error c4err = {};
+        c4collection = c4db_createCollection(_c4db, spec, &c4err);
+        if (!c4collection) {
+            CBLWarn(Database, @"%@ Failed to create collection: %@.%@ (%d/%d)",
+                    self, scopeName, name, c4err.domain, c4err.code);
+            convertError(c4err, error);
+            return nil;
+        }
+        CBLLogVerbose(Database, @"%@ Created collection %@.%@", self, scopeName, name);
+    }
     
-    CBLScope* s = [[CBLScope alloc] initWithName: scope ?: kCBLDefaultScopeName error: error];
-    return [[CBLCollection alloc] initWithName: name scope: s error: error];
+    return [[CBLCollection alloc] initWithDB: self c4collection: c4collection];
 }
 
 - (nullable CBLCollection*) collectionWithName: (NSString*)name
-                                scope: (nullable NSString*)scope
-                                error: (NSError**)error {
+                                         scope: (nullable NSString*)scope
+                                         error: (NSError**)error {
+    NSString* scopeName = scope ?: kCBLDefaultScopeName;
+    CBLStringBytes cName(name);
+    CBLStringBytes sName(scopeName);
+    C4CollectionSpec spec = { .name = cName, .scope = sName };
     
-    // TODO: add implementation
+    __block C4Collection* c;
+    CBL_LOCK(self) {
+        if (![self mustBeOpen: error])
+            return nil;
+        
+        C4Error c4err = {};
+        c = c4db_getCollection(_c4db, spec, &c4err);
+        if (c4err.code != 0) {
+            CBLWarn(Database, @"%@ Failed to get collection: %@.%@ (%d/%d)",
+                    self, scopeName, name, c4err.domain, c4err.code);
+            convertError(c4err, error);
+            return nil;
+        }
+        
+        if (!c)
+            return nil;
+    }
     
-    CBLScope* s = [[CBLScope alloc] initWithName: scope ?: kCBLDefaultScopeName error: nil];
-    return [[CBLCollection alloc] initWithName: name scope: s error: nil];
+    return [[CBLCollection alloc] initWithDB: self c4collection: c];
```

Comment:
Just notice this, should be returned inside the lock to be able to retain `c` before the other thread could cause `c` to be freed on the other thread (remember that `c` is cached in LiteCore and could be released anytime).